core/templates/pages/community-dashboard-page/services/contribution-opportunities.service.ts

```diff
@@ -17,73 +17,108 @@
-        translationOpportunitiesCursor = nextCursor;
-        moreTranslationOpportunitiesAvailable = more;
-        successCallback(opportunities, more);
+@Injectable({
+  providedIn: 'root'
+})
+export class ContributionOpportunitiesService {
+  skillOpportunitiesCursor = null;
+  translationOpportunitiesCursor = null;
+  voiceoverOpportunitiesCursor = null;
+  moreSkillOpportunitiesAvailable = true;
+  moreTranslationOpportunitiesAvailable = true;
+  moreVoiceoverOpportunitiesAvailable = true;
+
+  constructor(
+    private contributionOpportunitiesBackendApiService:
+      ContributionOpportunitiesBackendApiService) { }
+
+  private _getSkillOpportunities(cursor: string,
+      successCallback: (opportunities: SkillOpportunity[],
+      more: boolean) => void): void {
+    this.contributionOpportunitiesBackendApiService.fetchSkillOpportunities(
+      cursor).then((value: IFetchOpportunitySuccessCallbackParams) => {
+      this.skillOpportunitiesCursor = value.nextCursor;
+      this.moreSkillOpportunitiesAvailable = value.more;
+      successCallback(value.opportunities, value.more);
+    });
+  }
+
+  private _getTranslationOpportunities(languageCode: string, cursor: string,
+      successCallback: (opportunities: SkillOpportunity[],
+      more: boolean) => void): void {
+    this.contributionOpportunitiesBackendApiService
+      .fetchTranslationOpportunities(languageCode, cursor)
+      .then((value: IFetchOpportunitySuccessCallbackParams) => {
+        this.translationOpportunitiesCursor = value.nextCursor;
+        this.moreTranslationOpportunitiesAvailable = value.more;
+        successCallback(value.opportunities, value.more);
       });
-    };
-    var _getVoiceoverOpportunities = function(
-        languageCode, cursor, successCallback) {
-      ContributionOpportunitiesBackendApiService.fetchVoiceoverOpportunities(
-        languageCode, cursor).then(function(opportunities, nextCursor, more) {
-        voiceoverOpportunitiesCursor = nextCursor;
-        moreVoiceoverOpportunitiesAvailable = more;
-        successCallback(opportunities, more);
+  }
+
+  private _getVoiceoverOpportunities(languageCode: string, cursor: string,
+      successCallback: (opportunities: SkillOpportunity[],
+      more: boolean) => void) {
+    this.contributionOpportunitiesBackendApiService.fetchVoiceoverOpportunities(
+      languageCode, cursor).then(
+      (value: IFetchOpportunitySuccessCallbackParams) => {
+        this.voiceoverOpportunitiesCursor = value.nextCursor;
+        this.moreVoiceoverOpportunitiesAvailable = value.more;
+        successCallback(value.opportunities, value.more);
       });
-    };
+  }
+
+  public getSkillOpportunities(successCallback:
+    (opportunities: SkillOpportunity[], more: boolean) => void): void {
+    this._getSkillOpportunities('', successCallback);
+  }
+
+  public getTranslationOpportunities(languageCode: string,
+      successCallback: (opportunities: SkillOpportunity[],
```

Comment:
I don't think here it will be a SkillOpportunity. Also, if it's wrong then are we expecting unit test to fail?

Reply:
@DubeySandeep So there'll be other opportunities in this case? As it was `any[]` so I tried to figure out how I can replace this as best as possible.

Reply:
Please trace back and check from where this value is coming,
Start with _getTranslationOpportunities check how/where the opportunities values are generated check the service/function which generates it and loop over this process and I think you can find it. Let me know if it doesn't works, I'll help! :)

Reply:
@DubeySandeep Thanks! I traced back and I saw this block of code which is being used as an argument here: `ContributionOpportunitiesService.getMoreTranslationOpportunities(
                TranslationLanguageService.getActiveLanguageCode(),
                updateWithNewOpportunities);`:
<img width="690" alt="Screenshot 2020-05-06 at 19 11 47" src="https://user-images.githubusercontent.com/11369764/81207262-9939ef80-8fcd-11ea-879b-bf7fa1c153fe.png">

Reply:
Have you looked into this file: `core/templates/domain/opportunity/ExplorationOpportunitySummaryObjectFactory.ts`, I think this the object we need here.

Reply:
@DubeySandeep This has been resolved too

Reply:
@DubeySandeep PTAL!